Add Jinja2 tests for testing types

dagwieers opened the pull request:
##### SUMMARY
So Jinja2 only has tests for testing string types, or iterable types,
which is quite limiting. If you need to test a boolean value in Jinja2
the ~only~ option is to test it is not a string and convert it to a string
and compare it to 'True' or 'False'.

So these tests are essential constructs in playbooks and templates.

This includes tests for:
- boolean
- true
- false
- list
- integer
- float

So you can simply do things like:

    {% if result.value is false %}

or

    {% if result.value is list %}

Just like you already can do

    {% if result.value is none %}

and

    {% if result.value is mapping %}

See the included integration tests for a complete explanation of what Jinja2 already can do, and how this is insufficient in some cases where this would be useful.

This is related to #37514 

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
Jinja2 core tests

##### ANSIBLE VERSION
v2.6

dagwieers commented:
I upstreamed these tests to the Jinja project too, see https://github.com/pallets/jinja/pull/824

mattclay commented on test/sanity/code-smell/no-tests-as-filters.py:
```diff
@@ -44,6 +44,8 @@
     'skip': 'skipped',
 }
 
+IGNORED_TESTS = ('list')
```
Future additions to this list will result in cleaner diffs if this is reformatted:

```python
IGNORED_TESTS = (
    'list',
)
```

mattclay replied:
Also, why does this test need to be ignored?

sivel replied:
I'm guessing because jinja2 provides a `list` filter: http://jinja.pocoo.org/docs/dev/templates/#list

sivel replied:
Although I'm concerned about adding both a filter and test by the same name.

dagwieers replied:
Well, `sequence` is already taken and `tuple` doesn't really describe what it is. But today `string` already exists both as a filter as well as a test. So I have no objections to do the same.

I proposed this upstream, so we may want to delay this until they have made a decision ?

sivel commented:
@jimi-c This adds some tests that shadow built in python functions such as `list` and `float`.  I know we renamed the `type` filter to `type_debug` because adding filters that shadowed python functions allowed `safe_eval` and mark those functions as safe.  Is it ok to do that with list and float?

webknjaz commented on lib/ansible/plugins/test/core.py:
```diff
@@ -66,6 +67,44 @@ def skipped(result):
+# NOTE: The existing Jinja2 'sequence' test matches strings and dictionaries
+def test_list(value):
+    ''' Return true if the object is a list or tuple '''
+    return isinstance(value, Sequence) and not isinstance(value, string_types)
```
JFYI
In the internal slack channel we came up with an idea: https://github.com/ansible/ansible/pull/39924
So it might get into devel before you merge this

dagwieers replied:
Well, I think we should be pushing this to Jinja rather than invent our own.
That is why I moved it upstream: pallets/jinja#824

PS I personally don't like the name "is_item_sequence".

webknjaz replied:
Honestly, I just failed to invent a better name

mattclay replied:
Having it upstream would be nice, but I doubt we can change our minimum required Jinja2 version anytime soon.

dagwieers replied:
@mattclay The idea was to backport it once it is accepted upstream (and the names are set in stone). If we do it now, we risk additional confusion.

webknjaz commented on lib/ansible/plugins/test/core.py:
```diff
@@ -66,6 +67,44 @@ def skipped(result):
+
+
+# NOTE: The existing Jinja2 'sequence' test matches strings and dictionaries
+def test_list(value):
```
Now that #39924 has been merged into devel you can:
```python
from functools import partial

from ansible.module_utils.common.collections import is_sequence

test_list = partial(is_sequence, include_strings=False)
```

dagwieers replied:
I am not impressed, sorry. I don't think this will make it easier to upstream. I don't need an Ansible-only solution.

webknjaz replied:
I'm not talking about upstream. This is about this PR, which currently doesn't need to have duplicate code, like one in devel.

dagwieers committed:
Add Jinja2 tests for testing types

So Jinja2 only has tests for testing string types, or iterable types,
which is quite limiting. If you need to test a boolean value in Jinja2
the only option is to test it is not a string and convert it to a string
and compare it to 'True' or 'False'.

So these tests are essential constructs in playbooks and templates.

This PR includes integration tests to validate Jinja2 behaviour.

mattclay commented on test/integration/targets/jinja2_tests/aliases:
```diff
@@ -0,0 +1 @@
+shippable/posix/ci/group1
```
shippable/posix/group1

webknjaz commented on lib/ansible/plugins/test/core.py:
```diff
@@ -78,14 +79,52 @@ def finished(result):
+
+
+# NOTE: The existing Jinja2 'sequence' test also matches strings and dictionaries
+def test_list(value):
```
This code is already available in `ansible.module_utils.common.collections.is_sequence` for broader use.
So please follow earlier suggestion to reuse it: https://github.com/ansible/ansible/pull/37517#discussion_r192379339

webknjaz reviewed (changes requested):
Let's not tolerate copy-paste driven development

bcoca commented:
So instead of duplicating tests to allow older versions of Jinja2 to work the same, but mask changes/implementation details on newer ones, we've opted to add a 'jinja2' collection with these kind of tests/filters that are missing from older versions.  temporarily at https://galaxy.ansible.com/sivel/jinja2
